packages/snaps-controllers/src/interface/utils.ts

```diff
@@ -1,102 +1,175 @@
+ * @param component - The input element.
  * @param form - The parent form name of the input.
  * @returns The input state.
  */
-export const constructFormInputState = (
-  state: InterfaceState,
-  component: Input,
+function constructFormInputState(
+  oldState: InterfaceState,
+  component: InputElement,
   form: string,
-) => {
-  const oldFormState = state[form] as FormState;
-  const oldInputState = oldFormState?.[component.name];
-  return component.value ?? oldInputState ?? null;
-};
+) {
+  const oldFormState = oldState[form] as FormState;
+  const oldInputState = oldFormState?.[component.props.name];
+  return component.props.value ?? oldInputState ?? null;
+}
 
 /**
- * Assert that the component name is unique in state.
+ * Get the input field from a field element.
  *
- * @param state - The interface state to verify against.
- * @param name - The component name to verify.
+ * @param element - The field element.
+ * @returns The input element.
  */
-export const assertNameIsUnique = (state: InterfaceState, name: string) => {
-  assert(
-    state[name] === undefined,
-    `Duplicate component names are not allowed, found multiple instances of: "${name}".`,
-  );
-};
+function getFieldInput(element: FieldElement) {
+  if (Array.isArray(element.props.children)) {
+    return element.props.children[0];
+  }
+
+  return element.props.children;
+}
+
+/**
+ * Construct the state for a form input.
+ *
+ * @param oldState - The previous state.
+ * @param component - The field element.
+ * @param form - The parent form name of the input.
+ * @param newState - The new state.
+ * @returns The input state.
+ */
+function constructFormState(
+  oldState: InterfaceState,
+  component: FieldElement | ButtonElement,
+  form: string,
+  newState: FormState,
+): FormState {
+  if (component.type === 'Button') {
+    return newState;
+  }
+
+  const input = getFieldInput(component);
+  assertNameIsUnique(newState, input.props.name);
+
+  newState[input.props.name] = constructFormInputState(oldState, input, form);
+
+  return newState;
+}
 
 /**
- * Construct the interface state for a given component tree while preserving values for matching stateful components in the old state.
+ * Construct the interface state for a given component tree.
  *
  * @param oldState - The previous state.
  * @param component - The UI component to construct state from.
  * @param newState - The state that is being constructed.
  * @returns The interface state of the passed component.
  */
-export const constructState = (
+export function constructState(
   oldState: InterfaceState,
-  component: Component,
+  component: JSXElement,
   newState: InterfaceState = {},
-): InterfaceState => {
-  const { type } = component;
-  if (type === NodeType.Panel) {
-    return component.children.reduce(
-      (acc, node) => constructState(oldState, node, acc),
+): InterfaceState {
+  if (component.type === 'Box') {
+    if (Array.isArray(component.props.children)) {
```

Comment:
Hmm, the notion of children being arrays and a singular element complicates this logic. Wondering if we could abstract that away instead of considering that edge-case every time

Reply:
We could always turn children into an array?

Reply:
Seems simpler yeah